### PR TITLE
fix #283 Add transformGroup to ParallelFlux

### LINKED ISSUE #283: Add .transformEach() in ParallelFlux to support Function<Publisher<U>, Publisher<V>>

### PR DESCRIPTION
**Pull request by simonbasle**
@dfeist as suggested by @smaldini, went the `GroupedFlux` way hence the renaming to `transformGroup`.

Btw @smaldini I wonder if `transform` prefix is correct, isn't it closer to a `compose`, through the use of `groups()` and `flatMap()`?

### CODE REVIEW

**Comment by smaldini** [comment]
@simonbasle you are right it is technically a composeGroup (or groupMap? )

**Comment by dfeist** [comment]
Maybe, but this a significantly different feature to the intial issue created.  I was requesting a `transform` or `transformRail` that takes a:

`Function<? super Flux<T>, ? extends Publisher<V>> transformer` and is compataible with `transfrorm` in Fluz.

**Comment by dfeist** [comment]
On second thoughts, this may be compatible with what i need.

I have an issue with naming though, in Flux `compose` means deferred transformation of Flux, and `transform` means immediate transformation of Flux/ParrelelFlux.  So it's not clear why this should be `compose` and not `transform`.

**Comment by smaldini** [comment]
@dfeist its compose because it effectively applies the transformation in a flatMap.

**Comment by dfeist** [comment]
Deleted my previous comments to make things clearer (sorry)..

This looks good, and i think `group` makes sense if it's already in use.  On the naming i don't see this as a `map`, at least from the user persective.  If it's `transform` or `compose` should be defined based on if the function is applied at assembly or subscribe time.  Their are use cases for both transform and compose and both exist in Flux.  My initial issue was requesting the `transfom` assembly-time version, but both make sense.

**Comment by dfeist** [comment]
@smaldini Yeah, i saw looking at impl.  Can a assembly-time transform be introduced too?

**Comment by simonbasle** [comment]
the fact that the function is applied through `flatMap` makes it per-subscription, and thus closer to `compose`.
I'm afraid it won't be possible to do a parallel version of `transform` (assembly time), because the rails don't exist per se at assembly time (they are created upon subscription)

**Comment by smaldini** [comment]
should be doable via `from(groups().transform( f -> f.transform(xxx)))`

**Comment by dfeist** [comment]
Ok, not a specific need for assembly time, just ideal if possible (and to match Flux). This is great as is though.. thanks.

**Comment by dfeist** [comment]
@smaldini Right, that should work  surely... Avoiding the flatMap would be ideal..

**Comment by simonbasle** [comment]
doesn't work, as `groups()` returns a `Flux<GroupedFlux<Integer, T>>`... as I said, rails are not really available at assembly time.

**Comment by dfeist** [comment]
@simonbasle 👍 Thanks for checking. Actually, looking at it the `flapMap` is not used per data signal, rather just during subscribe.

I also checked this works with my use case where I need to compose, things like:
```
class MyStreamProcessor implements Function<Publisher<Integer>, Publisher<Integer>>
```

**Comment by simonbasle** [comment]
@dfeist great news, let's get this in then ;)

**Comment by smaldini** [comment]
I'm still preferring groupMap but if it doesn't rain through it doesn't (and would both mirror flatMap which it is, group and map).

**Comment by simonbasle** [comment]
(rebased on master, operator naming is still up in the air)

**Comment by dfeist** [comment]
@smaldini @simonbasle IMO naming depends if you name it based on what it's doing in implementation, or based on what it'd be used for, you could argue both ways.

From implementation perspective i'd agree, it's doing a groupFlatMap, but from the perspective of the user, i think they'd typiclly use this in order to do a `Flux.transform` on a parallel stream, more than thinking about the fact they are doing a flapMap on a group.  Isn't parallel attempting to hide the details of whats going on beyond and alllow it to be used like Flux?

**Comment by smaldini** [comment]
@dfeist it cannot be transform it both differs technically (it is deferred) and semantically (it is per rail). The closest as in Flux to this definition is compose and then come combinatory operators. We might stick with composeGroup then.

**Comment by dfeist** [comment]
@smaldini Sorry I mixed up. I meant to say `compose` in that last comment.

IMO user should be able to do the following with very similair semantics and be able to use `publisherFunction` in both scenarios in the same way:
```
flux.compose(publisherFunction).
flux.paralell(n).composeXX(publisherFunction).sequencial().
```

The only reason we started talking about `composeRail`, `composeGroup` etc, was because `compose`, (or transform couldn't be used directly because technically they would need a function returning a ParallelFlux and not an arbitrary Publisher.

`composeGroup` is good, as long as there is some mention in JavaDoc about that each rail processes a _group_ which is a subset of original flux. If not something should be added to this effect and then this is good to go IMO.

File: src/main/java/reactor/core/publisher/ParallelFlux.java [hunk 1]
```diff
@@ -947,6 +947,23 @@ public void subscribe(Consumer<? super T> onNext, Consumer<? super Throwable>
 	}
 
 	/**
+	 * Allows composing operators off the 'rails', as individual {@link GroupedFlux} instances keyed by
```
Comment (simonbasle) [inline src/main/java/reactor/core/publisher/ParallelFlux.java]: @dfeist does that seem explicit enough about the operator working on groups?

Comment (dfeist) [inline src/main/java/reactor/core/publisher/ParallelFlux.java]: Yep perfect, although I'd also mention/introduce the concept of each rail processing a group in the class-level Javadoc too (if that's not too picky :-) )

**Pull request merged**